bin/controller.ts

```diff
@@ -7,40 +7,57 @@ import {
 	Client,
 	createSSHConnection,
 	GenericClient,
-	NormalJob,
 	HeapJobOrderer,
-	Job,
 } from '../src'
+import { interpretArgv } from '../src/Controller/CommandLineController'
+import { makefileToJobTree } from '../src/Controller/MakefileToJobTree'
 
 const zeroconf = bonjour()
-
-const job3: Job = new NormalJob('third')
-const job4: Job = new NormalJob('fourth')
-const job2: Job = new NormalJob('second', new Set([job3]))
-const job1: Job = new NormalJob('first', new Set([job4, job2]))
-const job5: Job = new NormalJob('fifth', new Set([job1]))
-
-const client: Client = new GenericClient(
-	createSSHConnection,
-	process,
-	new HeapJobOrderer([job5]),
-)
-client.on('error', console.error)
-client.on('progress', (job, result) => console.log(job, result.status))
-client.on('done', (success): void => {
-	console.log('Finished')
-	process.exitCode = +!success
-	zeroconf.destroy()
-})
-
-process.on('SIGINT', () => {
-	console.log('\nClosing Daemons\n')
-
-	client.quit()
-
-	console.log('Exit Process\n')
-})
-
-const browser = supplyClient(zeroconf, client)
-browser.on('up', console.info)
-browser.on('down', console.info)
+const cliArgs = interpretArgv(process.argv.slice(2))
+
+if (cliArgs.cleanExit) {
+	process.exit(0)
+}
+
+// Exit on invalid arguments. Do this *after* checking for a clean exit, because
+// we want `--help` messages to take priority over invalid arguments.
+if (cliArgs.invalidArguments) {
+	console.error(
+		new Error(`Invalid arguments supplied: ${process.argv.slice(2)}.`),
+	)
+	process.exit(1)
+}
+
+makefileToJobTree({ filePath: cliArgs.makefile, targets: cliArgs.targets })
+	.then((rootJobs) => {
```

Comment:
Move this block into a separate function so you can make it `async`. That way you can use `await makefileToJobTree()` and make the control flow a lot easier to understand.